### PR TITLE
Add external OpenID login

### PR DESCRIPTION
**Pull request by t-aleksander**
## 📖 Description

Adds the ability to login using an external openid provider.

#### Documentation ###

- [x] If testing requires changes in the environment or deployment, please **update the documentation** (https://defguard.gitbook.io) first and **attach the link to the documentation** section in this pool request
- [x] I have commented on my code, particularly in hard-to-understand areas

#### Testing ### 

- [x] I have prepared end-to-end tests for all new functionalities
- [x] I have performed end-to-end tests manually and they work
- [x] New and existing unit tests pass locally with my changes

#### Deployment ###

- [x] If deployment is affected I have made corresponding/required changes to [deployment](https://github.com/defguard/deployment) (Docker, Kubernetes, one-line install)

#### Testing ### 

- [ ] I have merged my changes before to dev and the dev checklist is done
- [ ] I have tested all functionalities on the dev instance and they work

#### Documentation ###

- [ ] I have made corresponding changes to the **user & admin documentation** and added new features documentation with screenshots for users/admins

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,5 +1,7 @@
 Copyright 2023 teonite ventures sp. z o.o. (teonite)
 
+Note: The following license applies to the entire repository except for the "enterprise" directory.
```
Comment (teon) [inline LICENSE:3]: NOTE: The following license applies to the entire repository except for all the contents in the "enterprise" directory in this repository.
NOTE: For the license of the enterprise directory, go to enterprise/LICENSE.

File: src/enterprise/LICENSE [hunk 2]
```diff
@@ -0,0 +1 @@
+Files in this directory ("enterprise") are not under the default repository license. Contact sales<at>defguard.net for further information.
```
Comment (teon) [inline src/enterprise/LICENSE:1]: For enterprise license and usage of this code & binary distribution with this code please contact by email: sales<at>defguard.net

File: src/enterprise/handlers/openid_login.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    Ok((StatusCode::FOUND, headers, cookies, private_cookies))
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_get_provider_metadata() {
+        let metadata = get_provider_metadata("https://accounts.google.com")
```
Comment (j-chmielewski) [inline src/enterprise/handlers/openid_login.rs:383]: Don't perform network queries in unit tests.

File: src/enterprise/handlers/openid_providers.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+#[derive(Debug, Deserialize, Serialize)]
+pub struct AddProviderData {
+    name: String,
+    base_url: String,
+    client_id: String,
+    client_secret: String,
+}
+
+impl AddProviderData {
+    pub fn new(name: String, base_url: String, client_id: String, client_secret: String) -> Self {
```
Comment (j-chmielewski) [inline src/enterprise/handlers/openid_providers.rs:21]: take &str instead of String

File: web/src/pages/auth/Login/components/OidcButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  if (hostname === 'accounts.google.com') {
+    return <GoogleButton url={url} />;
+  } else if (hostname === 'login.microsoftonline.com') {
+    return <MicrosoftButton url={url} />;
+  } else {
+    return <CustomButton url={url} />;
+  }
+};
+
+const GoogleButton = ({ url }: { url: string }) => {
```
Comment (j-chmielewski) [inline web/src/pages/auth/Login/components/OidcButtons.tsx:20]: How does this work with any other oidc provider?

File: web/src/pages/users/UserProfile/UserProfile.tsx [hunk 6]
```diff
@@ -182,7 +182,10 @@ const EditModeControls = () => {
           size={ButtonSize.SMALL}
           styleVariant={ButtonStyleVariant.SAVE}
           icon={<IconCheckmarkWhite />}
-          onClick={() => submitSubject.next()}
+          onClick={() => {
+            submitSubject.next();
+            console.log('submitSubject.next();');
```
Comment (j-chmielewski) [inline web/src/pages/users/UserProfile/UserProfile.tsx:187]: Delet dis

File: src/lib.rs [hunk 7]
```diff
@@ -397,7 +402,12 @@ pub fn build_webapp(
             .route("/webhook/:id", delete(delete_webhook))
             .route("/webhook/:id", post(change_enabled))
             // ldap
-            .route("/ldap/test", get(test_ldap_settings)),
+            .route("/ldap/test", get(test_ldap_settings))
+            // OIDC login
+            .route("/openid/provider", get(get_current_openid_provider))
+            .route("/openid/provider", post(add_openid_provider))
+            .route("/openid/callback", get(auth_callback))
+            .route("/openid/auth_info", get(get_auth_info)),
```
Comment (j-chmielewski) [inline src/lib.rs:410]: `delete_openid_provider`, `modify_openid_provider`

**Review by j-chmielewski — CHANGES_REQUESTED** [review]

**Pull request merged**